gtsfm/averaging/rotation/shonan.py

```diff
@@ -56,23 +66,65 @@ def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
 
         for (i1, i2), i2Ri1 in i2Ri1_dict.items():
             if i2Ri1 is not None:
-                between_factors.append(
-                    BetweenFactorPose3(
-                        i2,
-                        i1,
-                        Pose3(
-                            i2Ri1,
-                            np.zeros(
-                                3,
-                            ),
-                        ),
-                        noise_model,
-                    )
-                )
+                # ignore translation during rotation averaging
+                i2Ti1 = Pose3(i2Ri1, np.zeros(3))
+                between_factors.append(BetweenFactorPose3(i2, i1, i2Ti1, noise_model))
 
         obj = ShonanAveraging3(between_factors, shonan_params)
 
         initial = obj.initializeRandomly()
         result_values, _ = obj.run(initial, self._p_min, self._p_max)
 
-        return [result_values.atRot3(i) for i in range(num_images)]
+        wRi_list_consecutive = [None] * num_connected_nodes
+        for i in range(num_connected_nodes):
+            if result_values.exists(i):
+                wRi_list_consecutive[i] = result_values.atRot3(i)
+
+        return wRi_list_consecutive
+
+    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
+        """Run the rotation averaging with arbitrary keys, where each key is a image/pose index.
+
+        Note: run() functions as a wrapper that re-orders keys to prepare a connected graph,
+        w/ N keys ordered [0,...,N-1].
+
+        Args:
+            num_images: number of images. Since we have one pose per image, it is also the number of poses.
+            i2Ri1_dict: relative rotations for each image pair-edge as dictionary (i1, i2): i2Ri1.
+
+        Returns:
+            Global rotations for each camera pose, i.e. wRi, as a list. The number of entries in the list is
+                `num_images`. The list may contain `None` where the global rotation could not be computed (either
+                underconstrained system or ill-constrained system), or where the camera pose had no valid observation
+                in the input to run().
+        """
+        connected_nodes = set()
+        for (i1, i2) in i2Ri1_dict.keys():
+            connected_nodes.add(i1)
+            connected_nodes.add(i2)
+
+        connected_nodes = sorted(list(connected_nodes))
```

Comment:
any reason to sort them?

Reply:
mainly just to preserve a monotonically increasing property, so no nodes got swapped out of relative order. Not strictly necessary, but I thought might convey to the reader that there is an implicit relative order.

Reply:
ok cool.

Reply:
ok cool.